File: Java/game/src/main/java/com/yan/game/service/impl/ResultServiceImpl.java
package com.yan.game.service.impl;

import com.yan.game.dao.GameDao;
import com.yan.game.dto.InfoDto;
import com.yan.game.entity.Level;
import com.yan.game.entity.Result;
import com.yan.game.entity.User;
import com.yan.game.exception.EntityAbsentException;
import com.yan.game.service.IResultService;
import java.util.Collections;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class ResultServiceImpl implements IResultService {
    private static final Logger logger = LoggerFactory.getLogger(ResultServiceImpl.class);
    private GameDao gameDao;

    public ResultServiceImpl(GameDao gameDao) {
        this.gameDao = gameDao;
    }

    @Override
    public List<InfoDto> getUserInfo(Long id) {
        List<Result> results = gameDao.findUserById(id)
                .map(User::getResults)
                .orElse(Collections.emptyList());
        return getInfo(results);
    }

    @Override
    public List<InfoDto> getLevelInfo(Long id) {
        List<Result> results = gameDao.findLevelById(id)
                .map(Level::getResults)
                .orElse(Collections.emptyList());
        return getInfo(results);
    }

    private List<InfoDto> getInfo(List<Result> results) {
        return results.stream()
                .map(this::resultsToDto)
                .sorted((infoDto, t1) -> {
                    if (infoDto.getResult() != t1.getResult()) {
                        return infoDto.getResult() < t1.getResult() ? 1 : -1;
                    } else {
                        return infoDto.getUserId() > t1.getUserId() ? 1 : -1;
                    }
                })
                .collect(Collectors.toList());
    }

    @Override
    public void setInfo(InfoDto infoDto) {
        User user = getUser(infoDto.getUserId());
        Level level = getLevel(infoDto.getLevelId());
        if(user == null || level == null)
            return;

        Result result = new Result();
        result.setUser(user);
        result.setLevel(level);
        result.setPoints(infoDto.getResult());
        gameDao.saveResult(result);
    }

    private User getUser(Long id) {
        User user = null;
        try {
            user = gameDao.findUserById(id)
                    .orElseThrow(() -> new EntityAbsentException("user with id: " + id));
        } catch (Exception ex) {
            logger.error(ex.getMessage());
        }
        return user;
    }

    private Level getLevel(Long id) {
        Level level = null;
        try {
            level = gameDao.findLevelById(id).
                    orElseThrow(() -> new EntityAbsentException("level with id: " + id));
        } catch (Exception ex) {
            logger.error(ex.getMessage());
        }
        return level;
    }

    private InfoDto resultsToDto(Result result) {
        InfoDto userInfoDto = new InfoDto();
        userInfoDto.setResult(result.getPoints());
        userInfoDto.setUserId(result.getUser().getId());
        userInfoDto.setLevelId(result.getLevel().getId());
        return userInfoDto;
    }
}

File: Kotlin/game/src/main/kotlin/com/yan/game/service/impl/ResultServiceImpl.kt
package com.yan.game.service.impl

import com.yan.game.dao.LevelRepository
import com.yan.game.entity.Result as Score
import com.yan.game.dao.ResultRepository
import com.yan.game.dao.UserRepository
import com.yan.game.dto.InfoDto
import com.yan.game.responses.FailureResponse
import com.yan.game.responses.Response
import com.yan.game.responses.SuccessResponse
import com.yan.game.service.ResultService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class ResultServiceImpl @Autowired constructor(
    private val resultRepository: ResultRepository,
    private val userRepository: UserRepository,
    private val levelRepository: LevelRepository
) : ResultService {

    override fun getUserInfo(userId: Int): List<InfoDto> {
        return resultRepository.findByUserId(userId)
            .map(::toDto)
    }

    override fun getLevelInfo(levelId: Int): List<InfoDto> {
        return resultRepository.findByLevelId(levelId)
            .map(::toDto)
    }

    override fun setInfo(infoDto: InfoDto): Response {
        if(!checkRequest(infoDto))
            return FailureResponse()
        val user = userRepository.findById(infoDto.userId!!)
        val level = levelRepository.findById(infoDto.levelId!!)
        val points = infoDto.points
        val result = Score(null, user.get(), level.get(), points ?: -1)
        resultRepository.saveAndFlush(result)
        return SuccessResponse()
    }

    private fun checkRequest(infoDto: InfoDto): Boolean {
        return infoDto.userId != null &&
                infoDto.levelId != null &&
                infoDto.points != null
    }

    fun toDto(result: Score): InfoDto {
        var dto = InfoDto()
        dto.userId = result.user.id
        dto.levelId = result.level.id
        dto.points = result.points
        return dto
    }
}